Update Ax to v1.0.0

schmoelder opened the pull request:
[Ax v1.0.0](https://github.com/facebook/Ax/releases/tag/1.0.0)  was currently released which introduced a new, unified Client / "Ask and Tell" API.
This PR updates CADET-Process to the new interface.

## Todo
- [x] Initial values @flo-schu 
- [x] CADET-Process / post processing @flo-schu 
- [x] Nonlinear-constraints @flo-schu 
- [x] Multi-objective
- [x] Running in batches (out of the box)
- [x] Restore from checkpoints @flo-schu 
- [ ] GPEI / NEHVI / qNParEGO @schmoelder 
- [ ] Early stopping (nice to have, probably later)

Note, a lot of the previously pre-configured Algorithms (e.g. GPEI) are not provided any more. As far as we understood, Ax analyses the problem at hand and tries to configure a suitable set of tools (which we still need to better understand). 

Users can build [custom Generators via the Modular BoTorch Interface](https://ax.dev/docs/tutorials/modular_botorch). We're not quite sure yet how to expose this.

Another alternative is to checkout the testing / documentation of Ax where some of the older algorithms are still configured:
- https://github.com/facebook/Ax/blob/28ad3d6c1a26db9862ac45aadf4f57958cd88cb3/ax/storage/botorch_modular_registry
- https://github.com/facebook/Ax/blob/28ad3d6c1a26db9862ac45aadf4f57958cd88cb3/ax/generators/torch/botorch_moo_defaults.py
- https://github.com/facebook/Ax/blob/28ad3d6c1a26db9862ac45aadf4f57958cd88cb3/ax/utils/testing/modeling_stubs.py
- https://github.com/facebook/Ax/blob/28ad3d6c1a26db9862ac45aadf4f57958cd88cb3/ax/generators/tests/test_botorch_moo_model.py#L106
- ...

flo-schu committed:
WIP: Dirty, refactor Ax to client API.

- Tests are not yet passing. But Ax runs and works for LinearConstraintsSingleObjective
- post_processing still needs re-write, which fails tests

flo-schu committed:
Fixed post processing.

Much simpler with the new Ax API

flo-schu commented:
Post processing is now running. It's much simpler with the new API, and I've implemented it so that order of objectives/nlcs and variables is strictly maintained. 

The `LinearConstraintsSooTestProblem2` is passing
The `LinearConstraintsSooTestProblem` is failing. Here I found a bound error towards the end of the optimization.

flo-schu committed:
Dirty: Implement start from initial values

flo-schu commented:
@schmoelder post processing and initial values are implemented. Checkpoints should be straightforward. Nonlinear constraints don't know yet, probably too. I will see what I can get done next week.

flo-schu committed:
Nonlinear constraints migrated.

- lambda functions are not allowed because their names cannot be parsed
Not passing tests:
- partial(LinearConstraintsSooTestProblem, transform="linear"),
- LinearConstraintsMooTestProblem (Pareto does not converge)

flo-schu commented:
@schmoelder Nonlinear constraints are also migrated.
Most tests are passing

flo-schu committed:
Refactor restart from checkpoint for Ax Client API

- create _complete_trials private method, to reuse throughout _run
- add inline comments
- some typing fixes
- minor refactoring
- improve test settings

flo-schu committed:
Catch ax error once, where no candidates can be generated.

Successive errors will still be raised.

flo-schu committed:
Fix: Re-enable additional optimizer tests

flo-schu commented:
Hi @schmoelder 

I'm done with my part. The Ax features are migrated ✅ 

GPEI tests pass except MOO Problems with restart from checkpoint. I'm not sure how to fix that. The issue is that the aborting callback seems to be called for each individual during the post_processing of the population. I'm pretty sure there is an easy solution for this, and before I get my hands unnecessarily dirty, I would like to hand it over to you.

In addition, I've left some inline comments and TODOs for you. Please have a look. 
Also the code is still not cleaned; there are commented parts of the code and disabled sections (early stopping). The final refactor I would like to leave to you.

I noticed that some other tests are not passing, because of out-of-bounds errors and initial conditions errors.

I think there is still a bit to do, but the GPEI feature is mostly complete.

schmoelder commented:
Hi @flo-schu, and happy new year! :)

thanks a lot for pushing this. I'm taking a look now.
